File: game/presentation/match_event_batch.lua
---@class MatchEventBatchModule
local match_event_batch = {}

---@param diffs RollbackEventDiff[]
---@return MatchEvent[]
function match_event_batch.surviving(diffs)
    ---@type table<string, RollbackWrappedMatchEvent>
    local frame_match_events = {}
    for _, diff in ipairs(diffs) do
        for _, event in ipairs(diff.revoked) do
            frame_match_events[event.id] = nil
        end
        for _, replacement in ipairs(diff.replaced) do
            frame_match_events[replacement.before.id] = nil
            local event = replacement.after
            if event.domain:sub(1, 6) == "match/" then
                ---@cast event RollbackWrappedMatchEvent
                frame_match_events[event.id] = event
            end
        end
        for _, event in ipairs(diff.added) do
            if event.domain:sub(1, 6) == "match/" then
                ---@cast event RollbackWrappedMatchEvent
                frame_match_events[event.id] = event
            end
        end
    end

    local event_ids = {}
    for id in pairs(frame_match_events) do
        event_ids[#event_ids + 1] = id
    end
    table.sort(event_ids)

    local events = {}
    for _, id in ipairs(event_ids) do
        events[#events + 1] = frame_match_events[id].payload
    end
    return events
end

return match_event_batch

File: ts/packages/presentation/src/match_event_batch.ts
// Reduces a sequence of rollback event diffs down to the match events that
// survived every correction, in deterministic (sorted-id) order. Purely a
// read of already-decided rollback output — it never decides which events
// are correct, only which ones are still standing after the fact.

/** The rollback-wrapped envelope shared by match/combat/lifecycle events. */
export interface RollbackWrappedEvent<TPayload = unknown> {
  readonly id: string;
  readonly tick: number;
  readonly domain: string;
  readonly ordinal: number;
  readonly payload: TPayload;
}

/** A match event's payload shape — owned by the (Rust) sim, kept minimal here. */
export interface MatchEvent {
  readonly kind: string;
}

export type RollbackWrappedMatchEvent = RollbackWrappedEvent<MatchEvent>;

export interface RollbackEventReplacement {
  readonly before: RollbackWrappedEvent;
  readonly after: RollbackWrappedEvent;
}

export interface RollbackEventDiff {
  readonly added: readonly RollbackWrappedEvent[];
  readonly revoked: readonly RollbackWrappedEvent[];
  readonly replaced: readonly RollbackEventReplacement[];
}

const MATCH_DOMAIN_PREFIX = "match/";

function surviving(diffs: readonly RollbackEventDiff[]): MatchEvent[] {
  const frameMatchEvents = new Map<string, RollbackWrappedMatchEvent>();
  for (const diff of diffs) {
    for (const revoked of diff.revoked) {
      frameMatchEvents.delete(revoked.id);
    }
    for (const replacement of diff.replaced) {
      frameMatchEvents.delete(replacement.before.id);
      const after = replacement.after;
      if (after.domain.startsWith(MATCH_DOMAIN_PREFIX)) {
        frameMatchEvents.set(after.id, after as RollbackWrappedMatchEvent);
      }
    }
    for (const added of diff.added) {
      if (added.domain.startsWith(MATCH_DOMAIN_PREFIX)) {
        frameMatchEvents.set(added.id, added as RollbackWrappedMatchEvent);
      }
    }
  }

  const eventIds = [...frameMatchEvents.keys()].sort();

  const events: MatchEvent[] = [];
  for (const id of eventIds) {
    const wrapped = frameMatchEvents.get(id);
    if (wrapped === undefined) {
      throw new Error("match event batch lost a tracked event id");
    }
    events.push(wrapped.payload);
  }
  return events;
}

export const matchEventBatch = { surviving };
